curvefs/src/client/rpcclient/metaserver_client.cpp

```diff
@@ -497,163 +497,185 @@ bool GroupInodeIdByPartition(
+        auto taskCtx = std::make_shared<TaskContext>(
+            MetaServerOpType::BatchGetInodeAttr, task, fsId, inodeId);
+        BatchGetInodeAttrExcutor excutor(
+            opt_, metaCache_, channelManager_, taskCtx);
+        auto ret = ConvertToMetaStatusCode(excutor.DoRPCTask());
+        if (ret != MetaStatusCode::OK) {
+            attr->clear();
+            return ret;
         }
     }
     return MetaStatusCode::OK;
 }
 
 MetaStatusCode MetaServerClientImpl::BatchGetXAttr(uint32_t fsId,
-    std::set<uint64_t> *inodeIds,
+    const std::set<uint64_t> &inodeIds,
     std::list<XAttr> *xattr) {
-    uint32_t limit = opt_.batchLimit;
-    // group inodeid by partition
-    std::unordered_map<uint32_t, std::list<uint64_t>> inodeGroups;
-    if (!GroupInodeIdByPartition(fsId, metaCache_, inodeIds, &inodeGroups)) {
+    // group inodeid by partition and batchlimit
+    std::vector<std::list<uint64_t>> inodeGroups;
+    if (!SplitRequestInodes(fsId, inodeIds, &inodeGroups)) {
         return MetaStatusCode::NOT_FOUND;
     }
 
     // send rpc
     for (const auto &it : inodeGroups) {
-        auto iter = it.second.begin();
-        while (iter != it.second.end()) {
-            uint64_t inodeId = *iter;
-            auto task = RPCTask {
-                metaserverClientMetric_->batchGetXattr.qps.count << 1;
-                LatencyUpdater updater(
-                    &metaserverClientMetric_->batchGetXattr.latency);
-                BatchGetXAttrRequest request;
-                BatchGetXAttrResponse response;
-                request.set_poolid(poolID);
-                request.set_copysetid(copysetID);
-                request.set_partitionid(partitionID);
-                request.set_fsid(fsId);
-                request.set_appliedindex(
-                    metaCache_->GetApplyIndex(
-                        CopysetGroupID(poolID, copysetID)));
-                uint32_t batchLimit = limit;
-                while (iter != it.second.end() && batchLimit > 0) {
-                    request.add_inodeid(*iter);
-                    iter++;
-                    batchLimit--;
-                }
-                curvefs::metaserver::MetaServerService_Stub stub(channel);
-                stub.BatchGetXAttr(cntl, &request, &response, nullptr);
-
-                if (cntl->Failed()) {
-                    metaserverClientMetric_->batchGetXattr.eps.count << 1;
-                    LOG(WARNING) << "BatchGetXAttr Failed, errorcode = "
-                                 << cntl->ErrorCode()
-                                 << ", error content:" << cntl->ErrorText()
-                                 << ", log id = " << cntl->log_id();
-                    return -cntl->ErrorCode();
-                }
+        uint64_t inodeId = *it.begin();
+        auto task = RPCTask {
+            metaserverClientMetric_->batchGetXattr.qps.count << 1;
+            LatencyUpdater updater(
+                &metaserverClientMetric_->batchGetXattr.latency);
```

Comment:
the overall process is still serial, can we send requests parallelly?

Reply:
Yes, add a TODO and improve this in later pr.